internal/pkg/service/stream/storage/node/readernode/sliceupload/operator.go

```diff
@@ -0,0 +1,363 @@
+		go func() {
+			defer wg.Done()
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					o.checkSlices(ctx, wg)
+				}
+			}
+		}()
+	}
+
+	return nil
+}
+
+func (o *operator) checkSlices(ctx context.Context, wg *sync.WaitGroup) {
+	o.logger.Debugf(ctx, "checking slices upload conditions")
+
+	o.slices.ForEach(func(_ model.SliceKey, data *sliceData) (stop bool) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			o.checkSlice(ctx, data)
+		}()
+		return false
+	})
+}
+
+func (o *operator) checkSlice(ctx context.Context, data *sliceData) {
+	// Prevent multiple checks of the same slice
+	if !data.Lock.TryLock() {
+		return
+	}
+	defer data.Lock.Unlock()
+
+	// No plugin existing for given slice to be uploaded
+	err := o.plugins.ImporterFor(data.Slice.StagingStorage.Provider)
+	if err != nil {
+		o.logger.Errorf(ctx, "importer for provider: %v does not exists", data.Slice.StagingStorage.Provider)
+		return
+	}
+
+	// Slice uploading is already in progress
+	if data.Uploading {
+		return
+	}
+
+	// Slice has been modified by some previous check, but we haven't received an updated version from etcd yet
+	if data.Processed {
+		return
+	}
+
+	if !data.Slice.Retryable.Allowed(o.clock.Now()) {
+		return
+	}
+
+	volume, err := o.volumes.Collection().Volume(data.Slice.SliceKey.VolumeID)
+	if err != nil {
+		o.logger.Errorf(ctx, "unable to upload slice. No volume found for key: %v", data.Slice.SliceKey.VolumeID)
+		return
+	}
+
+	switch data.Slice.State {
+	case model.SliceUploading:
+		o.uploadSlice(ctx, volume, data)
+	default:
+		// nop
+	}
+}
+
+func (o *operator) uploadSlice(ctx context.Context, volume *diskreader.Volume, data *sliceData) {
+	// Empty slice does not need to be uploaded to staging. Just mark as uploaded
+	if data.Slice.LocalStorage.IsEmpty {
+		// Update the entity, the ctx may be cancelled
+		dbCtx, dbCancel := context.WithTimeout(context.WithoutCancel(ctx), 30*time.Second)
+		defer dbCancel()
+
+		err := o.storage.Slice().SwitchToUploaded(data.Slice.SliceKey, o.clock.Now()).Do(dbCtx).Err()
+		if err != nil {
+			o.logger.Errorf(dbCtx, "cannot switch slice to the uploaded state: %v", err)
+
+			// Increment retry delay
+			err = o.storage.Slice().IncrementRetryAttempt(data.Slice.SliceKey, o.clock.Now(), err.Error()).Do(ctx).Err()
+			if err != nil {
+				o.logger.Errorf(ctx, "cannot increment slice retry: %v", err)
+				return
+			}
+		}
+		return
+	}
+
+	// Get slice statistics
+	stats, err := o.statistics.SliceStats(ctx, data.Slice.SliceKey)
+	if err != nil {
+		o.logger.Errorf(ctx, "cannot get slice statistics: %s", err)
+		return
+	}
+
+	if !data.Uploading {
+		o.logger.Infof(ctx, `uploading slice %q`, data.Slice.SliceKey)
+		// Use plugin system to upload slice to stagin storage. Set is an in-progress upload
+		uploadCtx, uploadCancel := context.WithTimeout(context.WithoutCancel(ctx), o.config.SliceUploadTimeout.Duration())
+		defer func() {
+			data.Uploading = false
```

Comment:
Isn't the `Processed` flag enough?

Reply:
Based on the discussion, it is enough, as we expect that 2 imports won't happen so often. However I leave it there as sort of in memory optimization (When service is restarted, it uploads the slices again).

Reply:
The more I think about it, the more I think it should be thrown away.

Intro:
- Martin created there `uploading` flag, it indicates sucessful upload (so `uploaded` is IMHO better name).
- The flag is used, if the upload has been successful, but we were not able to change the slice entity state in etcd.
- So the upload is skipped next time - we try only to switch the entity state.
- I don't know, why it is set to the `true` before, not after the upload.

I suggest to delete it:
- So a regular retry mechanism with exponential backoff will be used.
   - Now it is retried each second? 
   - Can't the problem get worse under some circumstances? 
- Re-upload is not problem, we can overwrite the key in the staging storage.
- The code will be simpler.
- It is an edge-case that the upload was successful but we cannot change the status in etcd.

@jachym-tousek-keboola please check it out too

Reply:
Exactly.